### PR TITLE
[Feature][Connector-V2][Milvus] Support Milvus source & sink

### PR DESCRIPTION
**Pull request by Thomas-HuWei**
### Purpose of this pull request

Support Milvus source & sink.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+     * prepare the commit, will be called before {@link #snapshotState(long checkpointId)}. If you
+     * need to use 2pc, you can return the commit info in this method, and receive the commit info
+     * in {@link SinkCommitter#commit(List)}. If this method failed (by throw exception), **Only**
+     * Spark engine will call {@link #abortPrepare()}
+     *
+     * @return the commit info need to commit
+     */
+    @Override
+    public Optional<MilvusCommitInfo> prepareCommit() throws IOException {
+        return Optional.empty();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java:105]: call `batchWriter.flush();`?


reference
https://github.com/apache/seatunnel/blob/dbdbdf015bceb90767787be465399c00b7baa167/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSinkWriter.java#L133

File: seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/catalog/MilvusCatalog.java [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+                        String.format("Create collection failed, err=%s", response.getMessage()),
+                        response.getException());
+            }
+        } catch (Exception e) {
+            throw new CatalogException(
+                    String.format("Failed creating collection %s", tablePath.getFullName()), e);
+        }
+    }
+
+    private FieldType convertToFieldType(Column column, PrimaryKey primaryKey) {
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/catalog/MilvusCatalog.java:269]: create `MilvusTypeConverter.java`

move `convertToFieldType` and `convertToDataType` into `MilvusTypeConverter`



reference
https://github.com/apache/seatunnel/blob/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/psql/PostgresTypeConverter.java

File: seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/batch/MilvusBufferBatchWriter.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        writeCount++;
+    }
+
+    @Override
+    public boolean needFlush() {
+        return this.writeCount >= this.batchSize;
+    }
+
+    @Override
+    public boolean flush() {
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/batch/MilvusBufferBatchWriter.java]:
```suggestion
    public synchronized boolean flush() {
```

**Review by hailin0 — COMMENTED** [review]
update this files:

https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties
https://github.com/apache/seatunnel/blob/dev/config/plugin_config


add maven dependency into this module

https://github.com/apache/seatunnel/blob/dbdbdf015bceb90767787be465399c00b7baa167/seatunnel-dist/pom.xml#L577

File: seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java (hunk 1 above)
Comment (Thomas-HuWei) [reply seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java:105]: Don't know it work, keep it current. Can you provider an interface to support batchWrite.

File: seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/catalog/MilvusCatalog.java (hunk 2 above)
Comment (Thomas-HuWei) [reply seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/catalog/MilvusCatalog.java:269]: I've MilvusConvertUtils, and have move convertToDataType to MilvusConvertUtils. convertToFieldType contains options, can't move.

File: seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java (hunk 1 above)
Comment (hailin0) [inline seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java]:
```suggestion
    public Optional<MilvusCommitInfo> prepareCommit() throws IOException {
        batchWriter.flush();
        return Optional.empty();
```

Comment (Thomas-HuWei) [reply seatunnel-connectors-v2/connector-milvus/src/main/java/org/apache/seatunnel/connectors/seatunnel/milvus/sink/MilvusSinkWriter.java]: Thank you. Done!

**Comment by hailin0** [comment]
LGTM
cc @Hisoka-X 
Waiting for CI passed

File: docs/en/connector-v2/source/Mivlus.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+## Source Options
+
+|    Name    |  Type  | Required | Default |                                        Description                                         |
+|------------|--------|----------|---------|--------------------------------------------------------------------------------------------|
+| url        | String | Yes      | -       | The URL to connect to Milvus or Zilliz Cloud.                                              |
+| token      | String | Yes      | -       | User:password                                                                              |
+| database   | String | Yes      | default | Read data from which database.                                                             |
+| collection | String | No       | -       | If set, will only read one collection, otherwise will read all collections under database. |
+
+## Task Example
```
Comment (hailin0) [inline docs/en/connector-v2/source/Mivlus.md:44]: add data-type-mapping

reference
https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/source/MySQL-CDC.md#data-type-mapping

File: docs/en/connector-v2/sink/Mivlus.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+| url                  | String  | Yes      | -                            | The URL to connect to Milvus or Zilliz Cloud.             |
+| token                | String  | Yes      | -                            | User:password                                             |
+| database             | String  | No       | -                            | Write data to which database, default is source database. |
+| schema_save_mode     | enum    | No       | CREATE_SCHEMA_WHEN_NOT_EXIST | Auto create table when table not exist.                   |
+| enable_auto_id       | boolean | No       | false                        | Primary key column enable autoId.                         |
+| enable_upsert        | boolean | No       | false                        | Upsert data not insert.                                   |
+| enable_dynamic_field | boolean | No       | true                         | Enable create table with dynamic field.                   |
+| batch_size           | int     | No       | 1000                         | Write batch size.                                         |
+
```
Comment (hailin0) [inline docs/en/connector-v2/sink/Mivlus.md:47]: add data-type-mapping

reference
https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/sink/Doris.md#data-type-mapping

**Comment by hailin0** [comment]
Please check ci error

image

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/TableSchema.java [hunk 6]
```diff
@@ -68,6 +72,10 @@ public static final class Builder {
 
         private final List<ConstraintKey> constraintKeys = new ArrayList<>();
 
+        private final List<VectorIndex> vectorIndexes = new ArrayList<>();
+
+        private Boolean enableDynamicField;
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/TableSchema.java]: Please add some comment to tell other developer what's used for? And how/when to use it?

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/VectorIndex.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+public class VectorIndex implements Serializable {
+
+    private String indexName;
+
+    private String fieldName;
+
+    private String indexType;
+
+    private String metricType;
+}
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/VectorIndex.java]: ditto

**Review by Hisoka-X — APPROVED** [review]
LGTM. Thanks @Thomas-HuWei

**Review by hailin0 — APPROVED** [review]
LGTM

**Pull request merged**